### PR TITLE
SOPS: Add support for HashiCorp Vault token-based authentication

### LINKED ISSUE #516: Support loading a secret with a HashiCorp Vault token to decrypt sops secrets.
We want to be able to pass a secret containing a `HashiCorp Vault` token to a `kustomization`. The Kustomization-controller would pass the token to the `Sops Decryptor` for decrypting secrets.

This should work like it does with age and pgp, e.g.
```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: my-secrets
  namespace: default
spec:
  interval: 5m
  path: "./"
  sourceRef:
    kind: GitRepository
    name: my-secrets
  decryption:
    provider: sops
    secretRef:
      name: sops-vault-token
```

### PR DESCRIPTION
**Pull request by souleb**
fixes #516 

If implemented, the kustomize controller will be able to retrieve a secret containing a VAULT TOKEN and use it to decrypt the sops encrypted master key. Which will then be used to decrypt the data key that will decrypt the final data.

The kustomize-controller retrieve the secret specified in the kustomization in the same namespace. We keep the behaviour set by age and pgp decryption. 

In case a token does not exist for a HashiCorp vault encrypted token, we fall back to the default server to try decrypting/encrypting the data. This will preserve the behaviour for customer relying on VAULT_TOKEN env var.

Signed-off-by: Soule BA <bah.soule@gmail.com>

### CODE REVIEW

File: controllers/kustomization_decryptor.go [hunk 1]
```diff
@@ -147,24 +148,31 @@ func (kd *KustomizeDecryptor) ImportKeys(ctx context.Context) error {
...
+				if err := os.WriteFile(keyPath, value, os.ModePerm); err != nil {
 					return fmt.Errorf("unable to write key to storage: %w", err)
 				}
 				if err := kd.gpgImport(keyPath); err != nil {
 					return err
 				}
 			case ".agekey":
-				ageIdentities = append(ageIdentities, string(file))
+				ageIdentities = append(ageIdentities, string(value))
+			case ".vault-token":
```
Comment (hiddeco) [inline controllers/kustomization_decryptor.go:167]: As `secret.Data` is a map, and the [iteration order of maps is undefined in Go](https://go.dev/blog/maps#iteration-order), this has a chance on picking a random key every time if multiple are defined.

I think this should be changed to either support multiple tokens (and collect them like the age identities above), use a fixed name (e.g. `token.vault`), or at least be stable sorted.

File: docs/spec/v1beta2/kustomization.md [hunk 2]
```diff
@@ -1001,6 +1001,51 @@ spec:
       name: sops-age
 ```
 
+### HashiCorp Vault
+
+Export the `VAULT_ADDR`  and `VAULT_TOKEN` enviromnet variables to your shell,
+then use `sops` to encrypt a kubernetes secret (see [HashiCorp Vault](https://www.vaultproject.io/) for more details on enabling the transit backend and [sops](https://github.com/mozilla/sops#encrypting-using-hashicorp-vault) for using the `sops` command):
```
Comment (hiddeco) [inline docs/spec/v1beta2/kustomization.md:1007]:
```suggestion
then use `sops` to encrypt a kubernetes secret (see [HashiCorp Vault](https://www.vaultproject.io/) for more details on enabling the transit backend and [sops](https://github.com/mozilla/sops#encrypting-using-hashicorp-vault)
for using the `sops` command):
```

File: controllers/kustomization_decryptor.go (hunk 1 above)
Comment (souleb) [reply controllers/kustomization_decryptor.go:167]: We are expecting one key with the `vault-token` extension to be in the secret, but you are right we never know what might happen. I'll go for the fixed key name as I don't think we should accept several tokens.

File: controllers/kustomization_decryptor.go [hunk 3]
```diff
@@ -45,13 +45,17 @@ import (
 	intkeyservice "github.com/fluxcd/kustomize-controller/internal/sops/keyservice"
 )
 
-const DecryptionProviderSOPS = "sops"
+const (
+	DecryptionProviderSOPS = "sops"
+	vaultTokenFileName     = "sops.vault-token"
```
Comment (hiddeco) [inline controllers/kustomization_decryptor.go:50]:
```suggestion
	DecryptionVaultTokenFileName = "sops.vault-token"
```

File: controllers/kustomization_decryptor.go [hunk 4]
```diff
@@ -147,24 +151,34 @@ func (kd *KustomizeDecryptor) ImportKeys(ctx context.Context) error {
...
 					return fmt.Errorf("unable to write key to storage: %w", err)
 				}
 				if err := kd.gpgImport(keyPath); err != nil {
 					return err
 				}
 			case ".agekey":
-				ageIdentities = append(ageIdentities, string(file))
+				ageIdentities = append(ageIdentities, string(value))
+			case ".vault-token":
+				// make sure we have the expected value
```
Comment (hiddeco) [inline controllers/kustomization_decryptor.go:171]: `// Make sure we have the absolute file name` maybe?

File: internal/sops/hcvault/keysource.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+	raw, err := cli.Logical().Write(fullPath, payload)
+	if err != nil {
+		return fmt.Errorf("the encryption to %s has failed: %w", fullPath, err)
+	}
+	if raw == nil || raw.Data == nil {
+		return fmt.Errorf("the transit backend %s is empty", fullPath)
+	}
+	encrypted, ok := raw.Data["ciphertext"]
+	if !ok {
+		return fmt.Errorf("there's not encrypted data")
```
Comment (hiddeco) [inline internal/sops/hcvault/keysource.go:145]:
```suggestion
		return fmt.Errorf("there's no encrypted data")
```

File: internal/sops/hcvault/keysource.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+	raw, err := cli.Logical().Write(fullPath, payload)
+	if err != nil {
+		return nil, fmt.Errorf("the decryption from %s has failed: %w", fullPath, err)
+	}
+	if raw == nil || raw.Data == nil {
+		return nil, fmt.Errorf("the transit backend %s is empty", fullPath)
+	}
+	decrypted, ok := raw.Data["plaintext"]
+	if !ok {
+		return nil, fmt.Errorf("there's no decrypted data")
```
Comment (hiddeco) [inline internal/sops/hcvault/keysource.go:181]:
```suggestion
		return nil, fmt.Errorf("there's no decrypted data")
```

File: internal/sops/hcvault/keysource.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+	if raw == nil || raw.Data == nil {
+		return fmt.Errorf("the transit backend %s is empty", fullPath)
+	}
+	encrypted, ok := raw.Data["ciphertext"]
+	if !ok {
+		return fmt.Errorf("there's not encrypted data")
+	}
+	encryptedKey, ok := encrypted.(string)
+	if !ok {
+		return fmt.Errorf("the ciphertext cannot be casted to string")
```
Comment (hiddeco) [inline internal/sops/hcvault/keysource.go:149]:
```suggestion
		return fmt.Errorf("the ciphertext cannot be cast to string")
```

File: internal/sops/hcvault/keysource.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+	if raw == nil || raw.Data == nil {
+		return nil, fmt.Errorf("the transit backend %s is empty", fullPath)
+	}
+	decrypted, ok := raw.Data["plaintext"]
+	if !ok {
+		return nil, fmt.Errorf("there's no decrypted data")
+	}
+	dataKey, ok := decrypted.(string)
+	if !ok {
+		return nil, fmt.Errorf("the plaintest cannot be casted to string")
```
Comment (hiddeco) [inline internal/sops/hcvault/keysource.go:185]:
```suggestion
		return nil, fmt.Errorf("the plaintext cannot be cast to string")
```

File: controllers/kustomization_decryptor.go (hunk 3 above)
Comment (souleb) [reply controllers/kustomization_decryptor.go:50]: Why do you want to export it?

**Comment by souleb** [comment]
@hiddeco I have checked the Azure Vault PR. We should probably make the `func NewServer(prompt bool, homeDir, vaultToken string, agePrivateKeys []string) keyservice.KeyServiceServer` accept options, if we are to keep adding parameters.

File: docs/spec/v1beta2/kustomization.md [hunk 9]
```diff
@@ -1001,6 +1001,53 @@ spec:
       name: sops-age
 ```
 
+### HashiCorp Vault
+
+Export the `VAULT_ADDR`  and `VAULT_TOKEN` enviromnet variables to your shell,
+then use `sops` to encrypt a kubernetes secret (see [HashiCorp Vault](https://www.vaultproject.io/docs/secrets/transit) for more details on enabling the transit backend and [sops](https://github.com/mozilla/sops#encrypting-using-hashicorp-vault)).
```
Comment (hiddeco) [inline docs/spec/v1beta2/kustomization.md:1007]:
```suggestion
then use `sops` to encrypt a kubernetes secret (see [HashiCorp Vault](https://www.vaultproject.io/docs/secrets/transit)
for more details on enabling the transit backend and [sops](https://github.com/mozilla/sops#encrypting-using-hashicorp-vault)).
```

File: controllers/kustomization_decryptor.go (hunk 3 above)
Comment (hiddeco) [reply controllers/kustomization_decryptor.go]: Because it's static value being checked on and thus arguably part of a contract.

**Review by hiddeco — APPROVED** [review]
No further comments from me, nice work @souleb 🙇

File: .github/workflows/e2e.yaml [hunk 10]
```diff
@@ -51,6 +51,12 @@ jobs:
         uses: fluxcd/pkg/actions/kubectl@main
         with:
           version: 1.21.2
+      - name: Setup SOPS
```
Comment (stefanprodan) [inline .github/workflows/e2e.yaml:54]: Please create an actions in pkg for sops based on this https://github.com/fluxcd/pkg/blob/main/actions/kustomize/action.yml

**Review by stefanprodan — CHANGES_REQUESTED** [review]

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @souleb 🏅

**Pull request merged**